expression/util.go

```diff
@@ -463,7 +462,12 @@ func ColumnSubstituteImpl(expr Expression, schema *Schema, newExprs []Expression
 					_ = append(append(append(tmpArgs, refExprArr.Result()[0:idx]...), refExprArr.Result()[idx+1:]...), newFuncExpr)
 					_, newColl := DeriveCollationFromExprs(v.GetCtx(), append(v.GetArgs(), newFuncExpr)...)
 					if coll == newColl {
-						changed = checkCollationStrictness(coll, newFuncExpr.GetType().GetCollate())
+						if newFuncExpr.GetType().GetCollate() == arg.GetType().GetCollate() && newFuncExpr.Coercibility() == arg.Coercibility() {
+							// It's safe to use the new expression, otherwise some cases in projection push-down will be wrong.
+							changed = true
+						} else {
+							changed = checkCollationStrictness(coll, newFuncExpr.GetType().GetCollate())
+						}
```

Comment:
why must make the substitute happen, I thought it was just an optimization, at least it should not get the wrong result if the optimization does not take effect. /cc @winoros